File: src/main/java/body/core/lifecycle/LifecycleAsyncImpl.java
package body.core.lifecycle;

import body.core.lifecycleState.LifecycleState;

import java.util.concurrent.CompletableFuture;
import java.util.function.Function;

public class LifecycleAsyncImpl<T> implements LifecycleAsync {

    private String state;
    private T obj;

    public static <H> LifecycleAsync create(H obj) {
        return new LifecycleAsyncImpl<H>(obj);
    }


    LifecycleAsyncImpl(T obj) {
        this.state = LifecycleState.NO_STATE;
        this.obj = obj;
    }

    public T getObject() {
        return this.obj;
    }

    public String getState() {
        return this.state;
    }

    public Boolean isInitialize() {
        return (this.state == LifecycleState.INITIALIZE);
    }

    public Boolean isPreDestroy() {
        return (this.state == LifecycleState.DESTROY);
    }

    public Boolean isPostConstructor() {
        return (this.state == LifecycleState.CONSTRUCTOR);
    }

    public Boolean isResume() {
        return (this.state == LifecycleState.RESUME);
    }

    public <H, T> CompletableFuture<T> initialize(Function<H, T> fun) {
        return null;
    }

    public <H, T> CompletableFuture<T> preDestroy(Function<H, T> fun) {
        return null;
    }

    public <H, T> CompletableFuture<T> postConstructor(Function<H, T> fun) {
        return null;
    }

    public <H, T> CompletableFuture<T> resume(Function<H, T> fun)  {
        return null;
    }

}

File: src/main/java/body/core/lifecycle/LifecycleAsyncImpl.kt
package body.core.lifecycle

import body.core.lifecycleState.LifecycleState

import java.util.concurrent.CompletableFuture
import java.util.function.Function

class LifecycleAsyncImpl<T> internal constructor(private val obj: T, private var state: String) : LifecycleAsync<T> {

    init {
        this.state = LifecycleState.CONSTRUCTOR
    }

    companion object {
        fun <H> create(obj: H): LifecycleAsync<H>  = LifecycleAsyncImpl(obj, LifecycleState.NO_STATE)
    }

    override fun getObject() = this.obj

    override fun getState() =  this.state

    override fun isInitialize() = (this.state === LifecycleState.INITIALIZE)

    override fun isPreDestroy() = (this.state === LifecycleState.DESTROY)

    override fun isPostConstructor() = (this.state === LifecycleState.CONSTRUCTOR)

    override fun isResume() = (this.state === LifecycleState.RESUME)

    override fun isNotState() = (this.state === LifecycleState.NO_STATE)



    override fun <H, T> initialize(`fun`: Function<H, T>): CompletableFuture<T>? {
        return null
    }

    override fun <H, T> preDestroy(`fun`: Function<H, T>): CompletableFuture<T>? {
        return null
    }

    override fun <H, T> postConstructor(`fun`: Function<H, T>): CompletableFuture<T>? {
        return null
    }

    override fun <H, T> resume(`fun`: Function<H, T>): CompletableFuture<T>? {
        return null
    }



}
